Detectors/TRD/calibration/src/TrackBasedCalib.cxx

```diff
@@ -77,6 +90,78 @@ void TrackBasedCalib::calculateAngResHistos()
   // mAngResHistos.print();
 }
 
+int TrackBasedCalib::filldEdx(gsl::span<const TrackTRD>& tracks, bool isTPCTRD)
+{
+  auto& params = TRDCalibParams::Instance();
+  int nTracksSuccess = 0;
+  for (const auto& trkIn : tracks) {
+
+    if (trkIn.getNtracklets() < params.nTrackletsMinGainCalib) {
+      continue;
+    }
+
+    if (trkIn.getP() < params.pMin || trkIn.getP() > params.pMax) {
+      continue;
+    }
+
+    auto id = trkIn.getRefGlobalTrackId();
+    float dEdxTPC;
+    if (isTPCTRD) {
+      dEdxTPC = mTracksTPC[id].getdEdx().dEdxTotTPC;
+    } else {
+      dEdxTPC = mTracksTPC[mTracksITSTPC[id].getRefTPC()].getdEdx().dEdxTotTPC;
+    }
+    if (dEdxTPC < params.dEdxTPCMin || dEdxTPC > params.dEdxTPCMax) {
+      continue;
+    }
+
+    if (std::isnan(trkIn.getSnp())) {
+      LOG(alarm) << "Track with invalid parameters found: " << trkIn.getRefGlobalTrackId();
+      continue;
+    }
+
+    auto trkWork = trkIn;
+
+    for (int iLayer = NLAYER - 1; iLayer >= 0; --iLayer) {
+      if (trkWork.getTrackletIndex(iLayer) == -1) {
+        continue;
+      }
+      if (mNoiseCalib && mNoiseCalib->isTrackletFromNoisyMCM(mTrackletsRaw[trkWork.getTrackletIndex(iLayer)])) {
+        // ignore tracklets which originate from noisy MCMs
+        continue;
+      }
+      if (propagateAndUpdate(trkWork, iLayer, false)) {
+        break;
+      }
+
+      float trkPhi = o2::math_utils::asin(trkWork.getSnp());
+      float trkLength = sqrt(1 + tan(trkPhi) * tan(trkPhi) + trkWork.getTgl() * trkWork.getTgl());
+
+      const auto& tracklet = mTrackletsRaw[trkWork.getTrackletIndex(iLayer)];
+
+      float Q0 = tracklet.getQ0();
+      float Q1 = tracklet.getQ1();
+      float Q2 = tracklet.getQ2();
+      // TODO: Should we also correct for local gain here?
+      float localGainCorr = 1.;
+      if (mLocalGain)
+        localGainCorr = mLocalGain->getValue(tracklet.getHCID() / 2, tracklet.getPadCol(), tracklet.getPadRow());
```

Comment:
@pachmay, @tdietel  should the gain calibration use the krypton calibration, it seemed to me that they were independent?

Reply:
the final tracklet charges used for PID should be corrected for chamber wise gain and local gain (krypton)

we could also correct the charges with the local gain before the chamber wise gain calibration, but I assume that there is not much difference